src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1978,183 +1987,205 @@ BuildJobTreeTaskList(Job *jobTree)
-													uniqueFragmentList);
-		int32 dataFetchTaskCount = list_length(dataFetchTaskList);
-		taskIdIndex += dataFetchTaskCount;
 
-		ExtractRangeTableRelationWalker((Node *) taskQuery, &taskRangeTableList);
-		UpdateRangeTableAlias(taskRangeTableList, fragmentCombination);
+	/* mark that we don't want the router planner to generate dummy hosts/queries */
+	replacePrunedQueryWithDummy = false;
 
-		/* transform the updated task query to a SQL query string */
-		sqlQueryString = makeStringInfo();
-		pg_get_query_def(taskQuery, sqlQueryString);
+	/*
+	 * Use router select planner to decide on whether we can push down the query
+	 * or not. If we can, we also rely on the side-effects that all RTEs have been
+	 * updated to point to the relevant nodes and selectPlacementList is determined.
+	 */
+	routerPlannable = RouterSelectQuery(copiedQuery, copiedRestrictionContext,
+										&selectPlacementList, &selectAnchorShardId,
+										&relationShardList, replacePrunedQueryWithDummy);
 
-		sqlTask = CreateBasicTask(jobId, taskIdIndex, SQL_TASK, sqlQueryString->data);
-		sqlTask->dependedTaskList = dataFetchTaskList;
+	if (!routerPlannable)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("cannot perform distributed planning for the given "
+							   "query"),
+						errdetail("Select query cannot be pushed down to the worker.")));
+	}
 
-		/* log the query string we generated */
-		ereport(DEBUG4, (errmsg("generated sql query for job " UINT64_FORMAT
-								" and task %d", sqlTask->jobId, sqlTask->taskId),
-						 errdetail("query string: \"%s\"", sqlQueryString->data)));
 
-		sqlTask->anchorShardId = AnchorShardId(fragmentCombination, anchorRangeTableId);
+	/* ensure that we do not send queries where select is pruned away completely */
+	if (list_length(selectPlacementList) == 0)
+	{
+		ereport(DEBUG2, (errmsg("Skipping target shard interval %ld since "
+								"SELECT query for it pruned away", shardId)));
 
-		taskIdIndex++;
-		sqlTaskList = lappend(sqlTaskList, sqlTask);
+		return NULL;
 	}
 
-	return sqlTaskList;
+
+	/*
+	 * Ands are made implicit during shard pruning, as predicate comparison and
+	 * refutation depend on it being so. We need to make them explicit again so
+	 * that the query string is generated as (...) AND (...) as opposed to
+	 * (...), (...).
+	 */
+	copiedQuery->jointree->quals =
+		(Node *) make_ands_explicit((List *) copiedQuery->jointree->quals);
+
+	/* and generate the full query string */
+	deparse_shard_query(copiedQuery, distributedTableId, shardInterval->shardId,
+						queryString);
+	ereport(DEBUG4, (errmsg("distributed statement: %s", queryString->data)));
+
+	subqueryTask = CreateBasicTask(jobId, taskIdIndex, SQL_TASK, queryString->data);
+	subqueryTask->dependedTaskList = NULL;
+	subqueryTask->anchorShardId = selectAnchorShardId;
+	subqueryTask->taskPlacementList = selectPlacementList;
+	subqueryTask->upsertQuery = upsertQuery;
```

Comment:
I would set this directly to `false` because we see what we are setting on this line differently from function calls.